dogma.go

```diff
@@ -385,20 +385,68 @@ func (c *Client) Push(ctx context.Context, projectName, repoName, baseRevision s
 	return c.content.push(ctx, projectName, repoName, baseRevision, commitMessage, changes)
 }
 
-// WatchFile awaits and returns the query result of the specified file since the specified last known revision.
-func (c *Client) WatchFile(ctx context.Context, projectName, repoName, lastKnownRevision string,
-	query *Query, timeout time.Duration) <-chan *WatchResult {
-	return c.watch.watchFile(ctx, projectName, repoName, lastKnownRevision, query, timeout)
+func (c *Client) watchWithWatcher(w *Watcher) (result <-chan WatchResult, closer func()) {
+	// setup watching channel
+	ch := make(chan WatchResult, DefaultChannelBuffer)
+	result = ch
+	w.Watch(func(value WatchResult) {
+		ch <- value
+	})
+
+	// setup closer
+	closer = func() {
+		w.Close()
+	}
+
+	// start watching
+	w.start()
+	return
+}
+
+// WatchFile watches on file changes. The watched result will be returned
+// through the returned channel. The api also provides manual closer to stop watching
+// and release underlying resources.
+// In short, watching will be stopped in case either context is cancelled or closer is
+// called.
+// Manually closing returned channel is unsafe and may cause sending on closed channel error.
+func (c *Client) WatchFile(ctx context.Context, projectName, repoName string,
```

Comment:
I'm wondering if it's better to remove the `ctx` or does not return the `closer` because they work in the same way to stop the watcher and we do not have to provide two ways for the same purpose. What do you think @linxGnu?

Reply:
Actually, I found `Watcher` interface is suitable more to internal use. It's not very convenient to user. Since user have to register himself a listener after got `Watcher`. So it's very easy for them to use directly with a channel, a closer to stop.
`ctx` here for user to add his own context. It's necessary too.
WDYT @minwoox

Reply:
Yeah, I agree with you on that the `Watcher` could be for internal use only because we provide a channel which a user can get the `WatchResult`.
Later, we might remove the `NewWatcher` API and listeners part so that the `Watcher` is not exposed to the user and the API is simplified.
So let's keep the API as it is for now. Thanks~!

Reply:
Yeah, I agree. I think we could remove it later on next PR. This PR is for api signature to make user easy to use, also simplify some of our internal process.